### PR TITLE
Visual Encoders (ViT, DeiT) Support Rollout

### PR DESCRIPTION
**Pull request by david-wei-01001**
# Description

Extended model_bridge of TransformerLens to support visual encoders (ViT, Deit)
Thoroughly tested:
unit test in ```tests/unit/model_bridge/supported_architectures/test_vit_adapter.py```
intergration test in ```tests/integration/model_bridge/test_vit_adapter.py```

can be tested as:

```bash
cd TransformerLens 
uv sync

uv run pytest tests/unit/model_bridge/supported_architectures/test_vit_adapter.py -v
uv run pytest tests/integration/model_bridge/test_vit_adapter.py -v 
```

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Screenshots
Please attach before and after screenshots of the change if applicable.

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by jlarson4** [comment]
Hi @david-wei-01001! Before I do a full review, would you mind pulling the latest changes from `dev` into your branch? We are moving from `transformers 5.8.1` to `transformers 5.13.0` in the next release, and that will have some effect on your code.

**Comment by david-wei-01001** [comment]
Resolved the format check
everything in dev is merged to this new PR

**Comment by jlarson4** [comment]
Thanks @david-wei-01001! I will do a full review within the next 24 hours. Sorry for messing with your branch while you were working on it. I saw that you'd merged `dev` right before I merged PR #1542 into `dev`, which caused a new set of conflicts, and didn't want to ask you to do the same process a second time.

**Comment by david-wei-01001** [comment]
> Thanks @david-wei-01001! I will do a full review within the next 24 hours. Sorry for messing with your branch while you were working on it. I saw that you'd merged `dev` right before I merged PR #1542 into `dev`, which caused a new set of conflicts, and didn't want to ask you to do the same process a second time.

Thank you very much! If you need anything from me, please let me know.

**Comment by david-wei-01001** [comment]
All checks pass now, ready to merge

File: transformer_lens/model_bridge/supported_architectures/vit.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    def prepare_loading(self, model_name: str, model_kwargs: dict) -> None:
+        """Propagate ViT/DeiT HF config attributes not covered by the generic mapper."""
+        hf_config = model_kwargs.get("config")
+        if hf_config is None:
+            return
+
+        head_dim = getattr(hf_config, "head_dim", None)
+        if head_dim is not None:
+            self.cfg.d_head = head_dim  # type: ignore[attr-defined]
```
Comment (jlarson4) [inline transformer_lens/model_bridge/supported_architectures/vit.py]: I removed this locally and everything passed cleanly, just want to confirm you're seeing an issue that requires it on your end? I'd like to remove if possible, we try not to `type: ignore` if we can avoid it.

File: tests/unit/model_bridge/supported_architectures/test_vit_adapter.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+`k_proj`/`v_proj`/`o_proj` directly (not nested `attention.query`/`attention.key`/
+`attention.value`/`output.dense`); and `ViTLayer.mlp` is already a flat `ViTMLP`
+with `fc1`/`fc2` (not `intermediate.dense`/`output.dense`). All the HF-path
+assertions below reflect that.
+
+NOT covered here (needs a real HF model + real forward pass — see the
+integration test at tests/integration/model_bridge/test_vit_adapter.py instead):
+- Numeric parity with HF
+- Hook firing / cache shapes
+- Whether cfg.n_ctx ends up correct (it currently does NOT — prepare_loading()
```
Comment (jlarson4) [inline tests/unit/model_bridge/supported_architectures/test_vit_adapter.py]: This docstring says `n_ctx` is never set and line 38 points at an `xfail` integration test that doesn't exist, but this PR does set `n_ctx`. Update the docstring and add the an assertion (`cfg.n_ctx == 197` for vit-base) to the integration file.

File: transformer_lens/model_bridge/supported_architectures/vit.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        # Derive n_ctx to avoid silent fallback to 2048 in generic logic
+        image_size = getattr(hf_config, "image_size", None)
+        patch_size = getattr(hf_config, "patch_size", None)
+        if image_size is not None and patch_size is not None:
+            num_patches = (image_size // patch_size) ** 2
+            # Standard ViT models add 1 cls_token to the sequence
+            self.cfg.n_ctx = num_patches + 1
```
Comment (jlarson4) [inline transformer_lens/model_bridge/supported_architectures/vit.py]: `n_ctx` derivation exists here and in `sources/transformers.py:160`. This copy raises `TypeError` on tuple/list `image_size` while the `sources/transformers.py` version is tuple-safe. Are both implementations necessary? If not, keep one, and make sure it is tuple-safe.

File: tests/integration/model_bridge/test_vit_adapter.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+from transformers import DeiTModel, ViTForImageClassification, ViTModel
+
+from transformer_lens.model_bridge.bridge import TransformerBridge
+from transformer_lens.model_bridge.generalized_components.vision_classifier_head import (
+    VisionClassifierHeadBridge,
+)
+
+MODEL_VIT_CLASSIFIER = "google/vit-base-patch16-224"
+MODEL_VIT_BARE = "google/vit-base-patch16-224-in21k"
+MODEL_DEIT_CLASSIFIER = "facebook/deit-small-patch16-224"
```
Comment (jlarson4) [inline tests/integration/model_bridge/test_vit_adapter.py:48]: `MODEL_DEIT_CLASSIFIER` is never used, and the "deit."-prefix classifier path has no real-model test, the fixture at 215-235 loads a bare `DeiTModel` instead. Wire it into a real `DeiTForImageClassification` test or drop it.

File: tests/integration/model_bridge/test_vit_adapter.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    def test_blocks_point_at_flat_layers_attribute_no_prefix(self, vit_bare_bridge):
+        assert vit_bare_bridge.adapter.component_mapping["blocks"].name == "layers"
+
+    def test_forward_does_not_crash(self, vit_bare_bridge):
+        pixel_values = _pixel_values()
+        with torch.no_grad():
+            vit_bare_bridge(pixel_values)  # must not raise
+
+    def test_forward_return_value_is_not_a_plain_tensor(self, vit_bare_bridge):
+        """KNOWN GAP — flagging, not necessarily desired behavior.
```
Comment (jlarson4) [inline tests/integration/model_bridge/test_vit_adapter.py]: Bare-encoder `return_type="logits"` returns a raw HF output object, codified here as a KNOWN-GAP test. Lets either return `last_hidden_state` or raise clearly. This also cites transformers 5.8.1, maybe 5.13.0 will provide a better solution?

File: transformer_lens/model_bridge/supported_architectures/vit.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+)
+from transformer_lens.model_bridge.generalized_components.vision_classifier_head import (
+    VisionClassifierHeadBridge,
+)
+from transformer_lens.model_bridge.generalized_components.vision_embeddings import (
+    VisionEmbeddingsBridge,
+)
+
+
+class ViTArchitectureAdapter(ArchitectureAdapter):
```
Comment (jlarson4) [inline transformer_lens/model_bridge/supported_architectures/vit.py:54]: `verify_models` default phases are text-shaped and can't run on a tokenizer-less vision model. Add `applicable_phases: list[int] = []` on the class as has been done for `falcon_h1.py`. I'll make sure to make an issue that we need to update our `verify_models` system to handle vision models as a future extension.

File: transformer_lens/model_bridge/generalized_components/vision_classifier_head.py
Comment (jlarson4) [inline transformer_lens/model_bridge/generalized_components/vision_classifier_head.py:1]: Make sure you're exporting `VisionEmbeddingsBridge` / `VisionClassifierHeadBridge` in `generalized_components/__init__.py` like their audio/CLIP siblings.

File: transformer_lens/model_bridge/bridge.py [hunk 7]
```diff
@@ -1760,9 +1760,10 @@ def forward(
                 (fragile across HF versions) or exception-based layer skipping (corrupts
                 model state). Raises NotImplementedError if a non-None value is passed.
             stop_at_layer: Layer to stop forward pass at
-            pixel_values: Optional image tensor for multimodal models (e.g., LLaVA, Gemma3).
+            pixel_values: Optional image tensor for multimodal models (e.g., LLaVA, Gemma3)
+                and vision models (eg. ViT, DeiT).
                 The tensor is passed directly to the underlying HuggingFace model.
-                Only valid when cfg.is_multimodal is True.
+                Only valid when cfg.is_multimodal is True or cfg.is_visual_mdoel is True.
```
Comment (jlarson4) [inline transformer_lens/model_bridge/bridge.py]: `is_visual_mdoel` typo, should be `is_visual_model`

**Review by jlarson4 — COMMENTED** [review]
Thank you for getting this all up to date with the latest state of the repo, I greatly appreciate this. The adapter looks great, just some small nits on testing & other small configuration things that can sometimes slip by the wayside. Once these are wrapped up I will get this merged and ready for the next release!

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 1 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py]: yes please remove. Thank you very much for catching it. I got a loooooot of problem testing because I have to support transformers 5.13.0 which completely rewrote ViT and DeiT as 5.8.0, so everything breaks

File: tests/unit/model_bridge/supported_architectures/test_vit_adapter.py (hunk 2 above)
Comment (david-wei-01001) [reply tests/unit/model_bridge/supported_architectures/test_vit_adapter.py]: thanks! Initially it fall back to test model's n_ctx but vision model should not have n_ctx, that's why I say it does not have n_ctx, but then I realize that vision model break inmages into 16 x 16 patches, and so 224 by 224 image is broken into 14 by 14 = 196 patches, adding one cls token that is 197 ctx in a similar sense

File: transformer_lens/model_bridge/bridge.py (hunk 7 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/bridge.py]: Thanks!

Comment (jlarson4) [reply transformer_lens/model_bridge/bridge.py]: You're welcome!

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 6 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py:54]: sure. Is this already resolved or I need to "Add applicable_phases: list[int] = [] on the class as has been done for falcon_h1.py"?

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 1 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py]: Done

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 3 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py]: Done

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 6 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py:54]: done

File: transformer_lens/model_bridge/bridge.py (hunk 7 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/bridge.py]: Done

File: transformer_lens/model_bridge/generalized_components/vision_classifier_head.py
Comment (david-wei-01001) [reply transformer_lens/model_bridge/generalized_components/vision_classifier_head.py:1]: Done

File: transformer_lens/model_bridge/supported_architectures/vit.py (hunk 3 above)
Comment (david-wei-01001) [reply transformer_lens/model_bridge/supported_architectures/vit.py]: kept only the tuple-safe one in sources/transformers.py

File: tests/integration/model_bridge/test_vit_adapter.py (hunk 4 above)
Comment (david-wei-01001) [reply tests/integration/model_bridge/test_vit_adapter.py:48]: Done

File: tests/unit/model_bridge/supported_architectures/test_vit_adapter.py (hunk 2 above)
Comment (jlarson4) [reply tests/unit/model_bridge/supported_architectures/test_vit_adapter.py]: Makes sense, thanks for the explanation!

File: tests/integration/model_bridge/test_vit_adapter.py (hunk 5 above)
Comment (david-wei-01001) [reply tests/integration/model_bridge/test_vit_adapter.py]: Resolved

File: tests/unit/model_bridge/supported_architectures/test_vit_adapter.py (hunk 2 above)
Comment (david-wei-01001) [reply tests/unit/model_bridge/supported_architectures/test_vit_adapter.py]: Resolved, prepare_loading() is removed so test no longer needed. n_ctx test migrated to intergration test.

**Comment by david-wei-01001** [comment]
All issues resolved, all formats checked. All tests passed, ready to merge.

**Comment by jlarson4** [comment]
Great work @david-wei-01001!

**Comment by david-wei-01001** [comment]
Thank you very much!
